tempesta_fw/http_sticky.c

```diff
@@ -0,0 +1,469 @@
+#include "cfg.h"
+#include "client.h"
+#include "http_msg.h"
+
+#define STICKY_NAME_MAXLEN	(32)
+#define STICKY_NAME_DEFAULT	"__tfw"
+#define STICKY_KEY_MAXLEN	(sizeof(((TfwClient *)0)->cookie.hmac))
+
+/**
+ * @name       - name of sticky cookie;
+ * @name_eq    - @name plus "=" to make some operations faster;
+ */
+typedef struct sticky {
+	TfwStr		name;
+	TfwStr		name_eq;
+	u_int		enabled : 1,
+			enforce : 1;
+} TfwCfgSticky;
+
+static TfwCfgSticky tfw_cfg_sticky;
+static struct crypto_shash *tfw_sticky_shash;
+static char tfw_sticky_key[STICKY_KEY_MAXLEN];
+
+static int
+tfw_http_sticky_send_302(TfwHttpMsg *hm)
+{
+	TfwConnection *conn = hm->conn;
+	TfwClient *client = (TfwClient *)hm->conn->peer;
+	TfwStr chunks[3], cookie = { 0 };
+	DEFINE_TFW_STR(s_eq, "=");
+	TfwHttpMsg resp;
+	char buf[sizeof(client->cookie.hmac) * 2];
+
+	tfw_http_prep_hexstring(buf, client->cookie.hmac,
+				sizeof(client->cookie.hmac));
+
+	memset(chunks, 0, sizeof(chunks));
+	chunks[0] = tfw_cfg_sticky.name;
+	chunks[1] = s_eq;
+	chunks[2].data = buf;
+	chunks[2].len = sizeof(client->cookie.hmac) * 2;
+
+	cookie.chunks = (struct TfwStr *)chunks;
+	cookie.len = chunks[0].len + chunks[1].len + chunks[2].len;
+	__TFW_STR_CHUNKN_SET(&cookie, 3);
+
+	if (tfw_http_prep_302(&resp, hm, &cookie))
+		return -1;
+	tfw_cli_conn_send(conn, (TfwMsg *)&resp);
+
+	return 0;
+}
+
+static int
+search_cookie(TfwPool *pool, const TfwStr *cookie, TfwStr *val)
+{
+	const char *const cstr = tfw_cfg_sticky.name_eq.data;
+	const unsigned int clen = tfw_cfg_sticky.name_eq.len;
+	TfwStr *chunk, *end, *next;
+	TfwStr tmp = { .flags = 0, };
+	unsigned int n = TFW_STR_CHUNKN(cookie);
+
+	BUG_ON(!TFW_STR_PLAIN(&tfw_cfg_sticky.name_eq));
+
+	/* Search cookie name. */
+	end = (TfwStr*)cookie->chunks + TFW_STR_CHUNKN(cookie);
+	for (chunk = ( TfwStr *)cookie->chunks; chunk != end; ++chunk, --n) {
+		if (chunk->flags & TFW_STR_NAME) {
+			/*
+			 * Create temporary compound string, starting
+			 * with this chunk.
+			 * We do not use it's overall length now,
+			 * so do not set it.
+			 */
+			tmp.chunks = (struct TfwStr *)chunk;
+			__TFW_STR_CHUNKN_SET(&tmp, n);
+			if (tfw_str_eq_cstr(&tmp, cstr, clen,
+					    TFW_STR_EQ_PREFIX))
+				break;
+		}
+	}
+	if (chunk == end)
+		return 0;
+
+	/* Search cookie value, starting with next chunk. */
+	for (++chunk; chunk != end; ++chunk)
+		if (chunk->flags & TFW_STR_VALUE)
+			break;
+	BUG_ON(chunk == end);
+
+	/* Check if value is plain string, just return it in this case. */
+	next = chunk + 1;
+	if (likely(next == end || *(char *)next->data == ';')) {
+		TFW_DBG3("%s: plain cookie value: %.*s\n", __func__,
+			 (int)chunk->len, (char *)chunk->data);
+		*val = *chunk;
+		return 1;
+	}
+
+	/* Add value chunks to out-string. */
+	TFW_DBG3("%s: compound cookie value found\n", __func__);
+	val->chunks = (struct TfwStr *)chunk;
+	TFW_STR_CHUNKN_ADD(val, 1);
+	val->len = chunk->len;
+	for (; chunk != end; ++chunk) {
+		if (*(char *)chunk->data == ';')
```

Comment:
Is the typecast necessary here?

I can't comment each and every occurrence of an unnecessary typecast. Please check all occurrences in places related to `TfwStr{}` member name changes.